meadow-client/src/Marlowe/Blockly.purs

```diff
@@ -0,0 +1,1144 @@
+  | ChoseObservationType
+  | ValueGEObservationType
+  | ValueGObservationType
+  | ValueLEObservationType
+  | ValueLObservationType
+  | ValueEqObservationType
+  | TrueObservationType
+  | FalseObservationType
+  -- values
+  | CurrentBlockType
+  | CommittedValueType
+  | ConstValueType
+  | NegValueType
+  | AddValueType
+  | SubValueType
+  | MulValueType
+  | DivValueType
+  | ModValueType
+  | FromChoiceValueType
+  | FromOracleValueType
+
+instance showBlockType :: Show BlockType where
+  show BaseContractType = "contract"
+  -- contracts
+  show NullContractType = "contract_null"
+  show CommitContractType = "contract_commit"
+  show PayContractType = "contract_pay"
+  show BothContractType = "contract_both"
+  show ChoiceContractType = "contract_choice"
+  show WhenContractType = "contract_when"
+  show WhileContractType = "contract_while"
+  show ScaleContractType = "contract_scale"
+  show LetContractType = "contract_let"
+  show UseContractType = "contract_use"
+  -- observations
+  show BelowTimeoutObservationType = "observation_below_timeout"
+  show AndObservationType = "observation_and"
+  show OrObservationType = "observation_or"
+  show NotObservationType = "observation_not"
+  show ChoseThisObservationType = "observation_chose_this"
+  show ChoseObservationType = "observation_chose"
+  show ValueGEObservationType = "observation_ge"
+  show ValueGObservationType = "observation_g"
+  show ValueLEObservationType = "observation_le"
+  show ValueLObservationType = "observation_l"
+  show ValueEqObservationType = "observation_eq"
+  show TrueObservationType = "observation_true"
+  show FalseObservationType = "observation_false"
+  -- values
+  show CurrentBlockType = "value_current_block"
+  show CommittedValueType = "value_committed"
+  show ConstValueType = "value_const"
+  show NegValueType = "value_neg"
+  show AddValueType = "value_add"
+  show SubValueType = "value_sub"
+  show MulValueType = "value_mul"
+  show DivValueType = "value_div"
+  show ModValueType = "value_mod"
+  show FromChoiceValueType = "value_from_choice"
+  show FromOracleValueType = "value_form_oracle"
+
+blockDefinitions :: Array BlockDefinition
+blockDefinitions =
+  [ baseContract -- this is a container for the entire contract and always exists
+  -- contracts
+  , nullContract
+  , commitContract
+  , payContract
+  , bothContract
+  , choiceContract
+  , whenContract
+  , whileContract
+  , scaleContract
+  , letContract
+  , useContract
+  -- observations
+  , belowTimeoutObservation
+  , andObservation
+  , orObservation
+  , notObservation
+  , choseThisObservation
+  , choseObservation
+  , valueGEObservation
+  , valueGObservation
+  , valueLEObservation
+  , valueLObservation
+  , valueEqObservation
+  , trueObservation
+  , falseObservation
+  -- values
+  , currentBlockValue
+  , committedValue
+  , constValue
+  , negValue
+  , addValue
+  , subValue
+  , mulValue
+  , divValue
+  , modValue
+  , fromChoiceValue
+  , fromOracleValue
+  ]
```

Comment:
I wonder if you could auto-generate a list of all `BlockType` values a with `Bounded` and/or `Enum` instance on `BlockType`. If you can, then there are a few variables in this file you can automatically derive. And several functions where you can get exhausiveness checking for free.

For instance, assuming PS behaves similarly to Haskell in this respect:

```
-- Free exhausiveness checking.
toDefinition :: BlockType -> BlockDefinition
toDefinition ...

-- Free exhausiveness checking.
mkGenFunction :: forall a. Show a => Generator -> BlockType -> (BlocklyState -> Block -> Either String a) -> Effect Unit
mkGenFunction ...

-- Free completeness.
blockDefinitions :: Array BlockDefinition
blockDefinitions = toDefinition $ bottom..top
```

Reply:
Even if the range trick doesn't work the same, `toDefinition` and `mkGenFunction` can still benefit from compiler-checks.

Reply:
done, see the latest commit, I managed to simplify things quite a bit